File: 2019/day2/opcode.py
#!/usr/bin/env python
#


def main():
    opcode_data = open("day2input.txt", "r").readline()
    opcode_data = map(int, opcode_data.split(","))

    # Edits from program
    opcode_data[1] = 12
    opcode_data[2] = 2

    opcode_data = run(opcode_data)
    return opcode_data[0]

def run(data):
    program_counter = 0
    
    while (True):
        opcode = data[program_counter]
        if (opcode == 99):
            break
        input_1 = data[data[program_counter + 1]]
        input_2 = data[data[program_counter + 2]]
        output_pos = data[program_counter + 3]
        output = 0

        print [opcode, input_1, input_2, output_pos]

        if (opcode == 1):
            output = input_1 + input_2
        elif (opcode == 2):
            output = input_1 * input_2
        else:
            raise AssertionError("invalid opcode")

        print(data)

        data[output_pos] = output
        program_counter += 4

    return data

if __name__ == "__main__":
    print main()

File: 2019/day2/opcode.dart
import "dart:io";

import "package:args/args.dart";

ArgResults _argResults;

void main(List<String> args) {
  final parser = ArgParser();

  _argResults = parser.parse(args);

  Future<String> input = new File(_argResults.rest[0]).readAsString();
  input.then((input_string) => stdout.writeln(
      findNounAndVerb(input_string.split(",").map(int.parse).toList())));
}

int findNounAndVerb(List<int> input) {
  List<int> data = input.toList(growable: false);
  for (int noun = 0; noun < 100; noun++) {
    for (int verb = 0; verb < 100; verb++) {
      List<int> output = run(data, noun, verb);
      if (output[0] == 19690720) {
        return 100 * noun + verb;
      }
    }
  }
  return -1;
}

List<int> run(List<int> input, [int noun, int verb]) {
  int programCounter = 0;
  List<int> data = input.toList(growable: false);
  data[1] = noun != null ? noun : data[1];
  data[2] = verb != null ? verb : data[2];

  while (true) {
    int opcode = data[programCounter];
    if (opcode == 99) break;
    int input1 = data[data[programCounter + 1]];
    int input2 = data[data[programCounter + 2]];
    int outputPos = data[programCounter + 3];
    int output = 0;

    if (opcode == 1) {
      output = input1 + input2;
    } else if (opcode == 2) {
      output = input1 * input2;
    } else {
      throw new ArgumentError("Unknown opcode: $opcode");
    }

    data[outputPos] = output;
    programCounter += 4;
  }

  return data;
}
